pkg/ottl/contexts/internal/ctxcontext/context.go

```diff
@@ -0,0 +1,306 @@
+
+package ctxcontext // import "github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl/contexts/internal/ctxcontext"
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	"go.opentelemetry.io/collector/client"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"google.golang.org/grpc/metadata"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl/contexts/internal/ctxerror"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl/contexts/internal/ctxutil"
+)
+
+const (
+	Name   = "context"
+	DocRef = "https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/pkg/ottl/contexts/ottlcontext"
+)
+
+func PathGetSetter[K any](path ottl.Path[K]) (ottl.GetSetter[K], error) {
+	switch path.Name() {
+	case "client":
+		return accessClient[K](path)
+	case "grpc":
+		return accessGRPC[K](path)
+	default:
+		return nil, ctxerror.New(path.Name(), path.String(), Name, DocRef)
+	}
+}
+
+func accessGRPC[K any](path ottl.Path[K]) (ottl.GetSetter[K], error) {
+	nextPath := path.Next()
+	if nextPath == nil {
+		return nil, ctxerror.New(path.Name(), path.String(), Name, DocRef)
+	}
+	switch nextPath.Name() {
+	case "metadata":
+		if nextPath.Keys() == nil {
+			return accessGRPCMetadataKeys[K](), nil
+		}
+		return accessGRPCMetadataKey[K](nextPath.Keys()), nil
+	default:
+		return nil, ctxerror.New(nextPath.Name(), nextPath.String(), Name, DocRef)
+	}
+}
+
+func accessClient[K any](path ottl.Path[K]) (ottl.GetSetter[K], error) {
+	nextPath := path.Next()
+	if nextPath == nil {
+		return nil, ctxerror.New(path.Name(), path.String(), Name, DocRef)
+	}
+	switch nextPath.Name() {
+	case "addr":
+		return accessClientAddr(nextPath)
+	case "auth":
+		return accessClientAuth(nextPath)
+	case "metadata":
+		return accessClientMetadata(nextPath)
+	default:
+		return nil, ctxerror.New(nextPath.Name(), nextPath.String(), Name, DocRef)
+	}
+}
+
+func accessClientMetadata[K any](path ottl.Path[K]) (ottl.GetSetter[K], error) {
+	nextPath := path.Next()
+	if nextPath != nil {
+		return nil, ctxerror.New(nextPath.Name(), nextPath.String(), Name, DocRef)
+	}
+	if path.Keys() == nil {
+		return accessClientMetadataKeys[K](), nil
+	}
+	return accessClientMetadataKey[K](path.Keys()), nil
+}
+
+func accessClientAddr[K any](path ottl.Path[K]) (ottl.GetSetter[K], error) {
+	nextPath := path.Next()
+	if nextPath != nil {
+		return nil, ctxerror.New(nextPath.Name(), nextPath.String(), Name, DocRef)
+	}
+	if path.Keys() != nil {
+		return nil, ctxerror.New(path.Name(), path.String(), Name, DocRef)
+	}
+	return ottl.StandardGetSetter[K]{
+		Getter: func(ctx context.Context, _ K) (any, error) {
+			cl := client.FromContext(ctx)
+			if cl.Addr == nil {
+				return nil, nil
+			}
+			return cl.Addr.String(), nil
+		},
+		Setter: func(_ context.Context, _ K, _ any) error {
+			return errors.New("cannot set value in context.client.addr")
```

Comment:
```suggestion
			return errors.New("context.client.addr is read-only and cannot be modified")
```

Just to make it clear why the set failed. If you agree, could you please also apply it to all other messages?

Reply:
I saw you applied this suggestion only to this setter, could you please do the same for all others paths? We can maybe create a const for that message so we don't have to repeat it multiple times: 

```go
const readOnlyPathErrMsg = "%q is read-only and cannot be modified"
return fmt.Errorf(readOnlyPathErrMsg, "context.client.addr")
```